zcash_client_sqlite/src/wallet/transparent.rs

```diff
@@ -468,13 +480,50 @@ pub(crate) fn put_received_transparent_utxo<P: consensus::Parameters>(
             output.mined_height(),
             address,
             receiving_account,
+            true,
         )
     } else {
         // The UTXO was not for any of our transparent addresses.
         Err(SqliteClientError::AddressNotRecognized(*address))
     }
 }
 
+/// Returns the vector of [`TransactionDataRequest`]s that represents the information needed by the
+/// wallet backend in order to be able to present a complete view of wallet history and memo data.
+pub(crate) fn transaction_data_requests<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+) -> Result<Vec<TransactionDataRequest>, SqliteClientError> {
+    // We cannot construct transaction data requests for the case where we cannot determine the
+    // height at which to begin, so we require that either the target height or mined height be
+    // set. In practice, we should not encounter entries in `transparent_spend_search_queue`
+    // because under ordinary circumstances, it is populated via a call from
+    // `decrypt_and_store_transaction` on ordinary mined transaction data retrieved from the chain.
+    let mut address_request_stmt = conn.prepare_cached(
+        "SELECT ssq.address, IFNULL(t.target_height, t.mined_height)
+         FROM transparent_spend_search_queue ssq
+         JOIN transactions t ON t.id_tx = ssq.transaction_id
+         WHERE t.target_height IS NOT NULL
+         OR t.mined_height IS NOT NULL",
+    )?;
+
+    let result = address_request_stmt
+        .query_and_then([], |row| {
+            let address = TransparentAddress::decode(params, &row.get::<_, String>(0)?)?;
+            let block_range_start = BlockHeight::from(row.get::<_, u32>(1)?);
+            Ok::<TransactionDataRequest, SqliteClientError>(
+                TransactionDataRequest::SpendsFromAddress {
+                    address,
+                    block_range_start,
+                    block_range_end: Some(block_range_start + DEFAULT_TX_EXPIRY_DELTA),
```

Comment:
Check for an off-by-one here. (Remember that the expiry height is the last block height at which the tx *can* be mined.)

Reply:
Remember that this is a heuristic upper bound; the expiry height of the transaction that created the output that we're looking for the spend of has no real bearing on when that output is actually spent; we don't know anything about the transaction that spends it (which is the transaction that we're looking for.)

As I've mentioned elsewhere, what we really need to be able to fix this correctly is to be able to query for spends of the outpoint, rather than just transactions related to the address it is sent to.

Something that just occurred to me is that there's another thing we should do here, which is explicitly check whether the outpoint is currently part of the UTXO set. If not, then we at least can definitely know that it's currently _not spendable_, which is the important thing.

Reply:
OK, but we might as well check for a potential off-by-one in the heuristic, anyway.

Reply:
The block range start will ordinarily be the target height for the transaction, so if we want this to be end-exclusive we should add one.